doc/design/zarr-support-2.md 

```diff
@@ -0,0 +1,94 @@
+# Abstract
+
+Zarr files are stored in their entirety in S3.
+Zarr files are represented by a single Asset through the API.
+A treehash is computed as the zarr file is uploaded.
+Each node in the treehash in the hash is stored in S3 to avoid bloating the DB.
+When published, the zarr file is copied to a separate location in the bucket so that changes to the draft cannot affect the published file.
+
+# Requirements
+
+1. Zarr files are stored in a "directory" in S3.
+1. Each zarr file corresponds to a single Asset.
+1. The CLI uses some kind of tree hashing scheme to compute a checksum for the entire zarr file.
+1. The API verifies the checksum _immediately_ after upload.
+1. The system can theoretically handle zarr files with ~1 million subfiles, each of size 64 * 64 * 64 bytes ~= 262 kilobytes.
+1. Zarr metadata must be mutable for zarr files in draft dandisets.
+
+# Implementation
+
+## API endpoints
+
+* POST /api/zarr_files/
+  Start a new zarr file.
+  Returns a zarr ID
+* GET /api/zarr_files/{zarr_id}/
+  Returns some information about the zarr file, like name, S3 path+url, and checksum
+* POST /api/zarr_files/{zarr_id}/batch/start/
+  Ask to upload a batch of subfiles as part of the zarr file.
+  No more batch uploads are permitted until this one is completed or aborted.
+  Requires a list of file paths and ETags (md5 checksums)
+  The file paths may include already uploaded files; this is how updates are done.
+  Returns a list of presigned upload URLs
+* POST /api/zarr_files/{zarr_id}/batch/complete/
+  Completes an upload of a batch of subfiles.
+  Fail if any of the checksums do not match.
+  Also, recursively update the tree hash for every parent node for each child.
+* DELETE /api/zarr_files/{zarr_id}/batch/
+  Cancels a batch upload.
+  Any files already uploaded are deleted.
+  A new batch upload can then be started.
+* DELETE /api/zarr_files/{zarr_id}/batch/delete/
+  Deletes subfiles from S3, and updates the treehash accordingly.
+  Requires a list of file paths
+
+* POST /api/dandisets/{...}/versions/{...}/assets/zarr/{zarr_id}/
```

Comment:
I don't like the idea of accepting a `/{blob_id}` or a `/{zarr_id}` in an endpoint because those two flavors of ID identify different (related) objects, and allowing them to be used interchangeably makes for a more confusing API. I have a personal preference towards multiple well-scope endpoints instead of a single, "clever" endpoint that can accept anything.